File: patches/fix_cowork_linux.py
#!/usr/bin/env python3
# @patch-target: app.asar.contents/.vite/build/index.js
# @patch-type: python
"""
Enable Cowork on Linux with runtime backend selection.

The claude-cowork-service daemon can run in either "native" mode (executes
Claude Code directly on the host; no VM) or "kvm" mode (boots a QEMU/KVM
guest using the Windows VM bundle artefacts). Each mode creates a different
Unix socket so Claude Desktop can auto-detect which one is running:

    native → $XDG_RUNTIME_DIR/cowork-vm-service.sock
    kvm    → $XDG_RUNTIME_DIR/cowork-kvm-service.sock

Runtime selection precedence (evaluated in the bundle preamble, below):
  1. Explicit COWORK_VM_BACKEND env var wins (values: "native" / "kvm",
     anything else falls through to native).
  2. Auto-detect: if the kvm socket exists, run in kvm mode.
  3. Fallback: native.

The choice is stored on globalThis.__coworkKvmMode. Patches gate behavior on
it via JS ternaries so a single patched bundle runs either backend based on
what the user has started.

Patches:
  0. Mode preamble — detects backend + sets globalThis.__coworkKvmMode.
  A. VM client loader — extend Li (win32 check) to include Linux so the
     TypeScript VM client loads instead of the macOS Swift module.
  B. Socket path — replace Windows named pipe with the Linux Unix socket
     (kvm name vs native name picked at runtime).
  C. Bundle config — two edits that cooperate:
       C1) inject ",linux:{x64:[]}" so native mode sees an empty file list
           (C$() vacuously true → z2e() sets status=Ready, no download),
       C2) alias linux→win32 at the two Xs.files[process.platform] lookup
           sites *only when kvm mode is active*, so kvm downloads the same
           VM artefacts Windows does (rootfs.vhdx, vmlinuz, initrd).
  D. pathToClaudeCodeExecutable — dynamic resolution on Linux.
  E. Error detection — extend "/usr/local/bin/claude" includes check to
     match the Linux paths too.
  F. present_files — allow host outputs dir paths on native backend.
  G. vmProcessId guards — remove the early-return guards that fire after
     idle teardown; the daemon handles mountPath/delete-permission without
     needing vmProcessId. Applied unconditionally (upstream bug affecting
     both backends).
  H. smol-bin copy — gate the "copy smol-bin.vhdx into session dir" step so
     it runs on Linux *only in kvm mode* (native has no VM to boot).

Patches always run on clean, freshly-extracted bundles. There are NO
"already patched" fast paths — a missing anchor is a hard failure so we
notice upstream churn immediately.

Usage: python3 fix_cowork_linux.py <path_to_index.js>
"""

import sys
import os
import re


EXPECTED_PATCHES = 11  # preamble + A + B + C1 + C2 + D + E + F + G-mount + G-delete + H


# Runtime mode selector JS snippet, loaded from the sibling file that both
# this script and the Nim port share. Keeping the source on disk as plain
# .js (rather than a Python string constant) means there's one canonical
# copy — no codegen step, no drift between Python and Nim implementations.
_HERE = os.path.dirname(os.path.abspath(__file__))
with open(os.path.join(_HERE, "../js/cowork_mode_preamble.js"), "rb") as _f:
    MODE_PREAMBLE_JS = _f.read()


def patch_cowork_linux(filepath):
    """Enable Cowork VM features on Linux with runtime native/kvm switching."""

    print("=== Patch: fix_cowork_linux ===")
    print(f"  Target: {filepath}")

    if not os.path.exists(filepath):
        print(f"  [FAIL] File not found: {filepath}")
        return False

    with open(filepath, "rb") as f:
        content = f.read()

    patches_applied = 0

    # ── Patch 0: Mode preamble at file start (right after "use strict";) ──
    #
    # Must run before any module-level const that contains a template
    # literal with ${globalThis.__coworkKvmMode?…} — sandbox_refs B/C wrap
    # substrings inside such templates, and those templates are constructed
    # as soon as their enclosing function/object initializer runs (which
    # can happen during module load, before app.on("ready") would fire).
    strict_marker = b'"use strict";'
    strict_idx = content.find(strict_marker)
    if strict_idx == 0:
        insertion_point = len(strict_marker)
        content = content[:insertion_point] + MODE_PREAMBLE_JS + content[insertion_point:]
        print('  [OK] 0 mode preamble: injected after "use strict"')
        patches_applied += 1
    else:
        print('  [FAIL] 0 mode preamble: "use strict"; not at file start')

    # ── Patch A: VM client loader — extend Li check to include Linux ───
    vm_client_pattern = rb'([\w$]+)\?([\w$]+)=(\{vm:[\w$]+\}):\2=\(await import\("@ant/claude-swift"\)\)\.default'

    def vm_client_replacement(m):
        li_var = m.group(1)
        ef_var = m.group(2)
        vm_obj = m.group(3)
        return b"(" + li_var + b'||process.platform==="linux")?' + ef_var + b"=" + vm_obj + b":" + ef_var + b'=(await import("@ant/claude-swift")).default'

    content, count_a = re.subn(vm_client_pattern, vm_client_replacement, content)
    if count_a >= 1:
        print(f"  [OK] A VM client loader: extended to Linux ({count_a} match)")
        patches_applied += 1
    else:
        print("  [FAIL] A VM client loader: 0 matches")

    # ── Patch B: Socket path — mode-aware Unix socket on Linux ──────
    pipe_path = b'"\\\\\\\\.\\\\pipe\\\\cowork-vm-service"'
    pipe_search = b"=" + pipe_path

    if pipe_search in content:
        idx = content.index(pipe_search)
        start = idx - 1
        while start >= 0 and (content[start : start + 1].isalnum() or content[start : start + 1] in (b"_", b"$")):
            start -= 1
        start += 1
        var_name = content[start:idx]

        replacement = (
            var_name
            + b'=process.platform==="linux"?'
            + b'(process.env.XDG_RUNTIME_DIR||"/tmp")+(globalThis.__coworkKvmMode?"/cowork-kvm-service.sock":"/cowork-vm-service.sock")'
            + b":"
            + pipe_path
        )
        content = content[:start] + replacement + content[idx + len(pipe_search) :]
        print(f"  [OK] B socket path: runtime-selected Unix socket on Linux (var={var_name.decode()})")
        patches_applied += 1
    else:
        print("  [FAIL] B socket path: pipe path not found")

    # ── Patch C1: Inject ",linux:{x64:[]}" into bundle files config ──
    win32_marker = b"win32:{"
    win32_idx = content.find(win32_marker)
    c1_ok = False
    if win32_idx >= 0:
        x64_marker = b",x64:["
        x64_idx = content.find(x64_marker, win32_idx)
        if x64_idx >= 0:
            array_start = x64_idx + len(x64_marker) - 1
            depth = 0
            pos = array_start
            while pos < len(content):
                if content[pos : pos + 1] == b"[":
                    depth += 1
                elif content[pos : pos + 1] == b"]":
                    depth -= 1
                    if depth == 0:
                        break
                pos += 1

            after_array = pos + 1
            if content[after_array : after_array + 1] == b"}":
                inject = b",linux:{x64:[]}"
                content = content[: after_array + 1] + inject + content[after_array + 1 :]
                print("  [OK] C1 bundle config: linux platform added (empty file list)")
                patches_applied += 1
                c1_ok = True
            else:
                print("  [FAIL] C1 bundle config: unexpected structure after x64 array")
        else:
            print("  [FAIL] C1 bundle config: x64 array not found in win32 block")
    else:
        print("  [FAIL] C1 bundle config: win32 block not found")

    # ── Patch C2: Alias linux→win32 at the Xs.files[…] lookup sites
    #             (runtime-gated on globalThis.__coworkKvmMode). ─────
    bundle_lookup_re = re.compile(
        rb"(const ([\w$]+)=)process\.platform"
        rb"(,[\w$]+=[\w$]+\(\);return [\w$]+\.files\[\2\])"
    )

    def bundle_lookup_replacement(m):
        return (
            m.group(1)
            + b'(globalThis.__coworkKvmMode?(process.platform==="linux"?"win32":process.platform):process.platform)'
            + m.group(3)
        )

    content, c2_count = bundle_lookup_re.subn(bundle_lookup_replacement, content)
    if c2_count >= 1:
        print(f"  [OK] C2 bundle lookup alias: linux→win32 when kvm mode ({c2_count} site(s))")
        patches_applied += 1
    else:
        print("  [FAIL] C2 bundle lookup alias: no matching sites found")

    # Silence unused variable warning on c1_ok — kept for readability.
    del c1_ok

    # ── Patch D: pathToClaudeCodeExecutable — dynamic on Linux ──────
    claude_path_old = b'pathToClaudeCodeExecutable:"/usr/local/bin/claude"'
    claude_path_new = (
        b"pathToClaudeCodeExecutable:"
        b'(()=>{if(process.platform!=="linux")return"/usr/local/bin/claude";'
        b'const fs=require("fs");'
        b'for(const p of["/usr/bin/claude",'
        b'(process.env.HOME||"")+"/.local/bin/claude",'
        b'"/usr/local/bin/claude"])'
        b"if(fs.existsSync(p))return p;"
        b'try{return require("child_process").execSync("which claude",{encoding:"utf-8"}).trim()}'
        b"catch(e){}"
        b'return"claude"})()'
    )

    if claude_path_old in content:
        content = content.replace(claude_path_old, claude_path_new, 1)
        print("  [OK] D Claude Code path: dynamic resolution on Linux")
        patches_applied += 1
    else:
        print("  [FAIL] D Claude Code path: pattern not found")

    # ── Patch E: Error detection — extend Linux paths ───────────────
    error_detect_pattern = rb'([\w$]+)(\.includes\("/usr/local/bin/claude"\))'

    def error_detect_replacement(m):
        var = m.group(1)
        return b"(" + var + b'.includes("/usr/local/bin/claude")||' + var + b'.includes("/usr/bin/claude")||' + var + b'.includes("/.local/bin/claude"))'

    content, error_count = re.subn(error_detect_pattern, error_detect_replacement, content, count=1)
    if error_count >= 1:
        print("  [OK] E error detection: extended for Linux paths")
        patches_applied += 1
    else:
        print("  [FAIL] E error detection: pattern not found")

    # ── Patch F: present_files — allow host outputs dir paths ───────
    present_files_old_pattern = (
        rb"for\(const\{file_path:([\w$]+),vmPath:([\w$]+)\}of ([\w$]+)\)\{"
        rb"if\(([\w$]+)\(\2,([\w$]+)\.vmProcessName\)\)continue;"
        rb"\(([\w$]+)\?([\w$]+)\(\2,\6\):null\)===null&&([\w$]+)\.push\(\1\)\}"
    )

    def present_files_replacement(m):
        f_var = m.group(1)
        p_var = m.group(2)
        l_var = m.group(3)
        scratchpad_fn = m.group(4)
        t_var = m.group(5)
        c_var = m.group(6)
        resolve_fn = m.group(7)
        u_var = m.group(8)
        return (
            b"for(const{file_path:" + f_var + b",vmPath:" + p_var + b"}of " + l_var + b"){"
            b"if(" + scratchpad_fn + b"(" + p_var + b"," + t_var + b".vmProcessName))continue;"
            b"(" + c_var + b"?" + resolve_fn + b"(" + p_var + b"," + c_var + b"):null)===null&&"
            b"(()=>{const _ho=" + t_var + b".getHostOutputsDir();"
            b"if(_ho&&(" + f_var + b"===_ho||" + f_var + b'.startsWith(_ho+"/")))return;' + u_var + b".push(" + f_var + b")})()}"
        )

    content, count_f = re.subn(present_files_old_pattern, present_files_replacement, content)
    if count_f >= 1:
        print(f"  [OK] F present_files: host outputs dir allowed ({count_f} match)")
        patches_applied += 1
    else:
        print("  [FAIL] F present_files: pattern not found")

    # ── Patch G-mount: remove mountFolderForSession vmProcessId guard ──
    mount_guard_re = re.compile(
        rb"if\s*\(\s*!\s*[a-zA-Z_$][\w$]*\s*\|\|\s*!\s*[a-zA-Z_$][\w$]*\s*\)"
        rb"\s*return\s*\{\s*ok\s*:\s*!\s*1\s*,\s*error\s*:\s*"
        rb'"Session VM process not available\. '
        rb'The session may not be fully initialized\."\s*\}\s*;?'
    )
    content, mount_n = mount_guard_re.subn(b"", content)
    if mount_n >= 1:
        print(f"  [OK] G-mount vmProcessId guard: removed ({mount_n} match)")
        patches_applied += 1
    else:
        print("  [FAIL] G-mount vmProcessId guard: pattern not found")

    # ── Patch G-delete: remove delete-permission tool vmProcessId guard ─
    delete_guard_re = re.compile(
        rb"if\s*\(\s*!\s*[a-zA-Z_$][\w$]*\s*\)\s*return\s*\{\s*content\s*:"
        rb"\s*\[\s*\{\s*type\s*:\s*\"text\"\s*,\s*text\s*:\s*"
        rb'"Session VM process not available\. '
        rb'The session may not be fully initialized\."\s*\}\s*\]\s*,'
        rb"\s*isError\s*:\s*!\s*0\s*\}\s*;?"
    )
    content, delete_n = delete_guard_re.subn(b"", content)
    if delete_n >= 1:
        print(f"  [OK] G-delete vmProcessId guard: removed ({delete_n} match)")
        patches_applied += 1
    else:
        print("  [FAIL] G-delete vmProcessId guard: pattern not found")

    # ── Patch H: smol-bin copy gate — runtime-gated on kvm mode ─────
    smol_bin_gate_re = re.compile(
        rb'if\(process\.platform==="win32"\)(\{const [\w$]+=[\w$]+\(\),'
        rb'[\w$]+=[\w$]+\.join\(process\.resourcesPath,`smol-bin\.)'
    )
    content, smol_n = smol_bin_gate_re.subn(
        rb'if(process.platform==="win32"||process.platform==="linux"&&globalThis.__coworkKvmMode)\1',
        content,
    )
    if smol_n >= 1:
        print(f"  [OK] H smol-bin copy gate: kvm-mode Linux opt-in ({smol_n} match)")
        patches_applied += 1
    else:
        print("  [FAIL] H smol-bin copy gate: win32 gate pattern not found")

    # ── Checks ───────────────────────────────────────────────────────
    if patches_applied < EXPECTED_PATCHES:
        print(f"  [FAIL] Only {patches_applied}/{EXPECTED_PATCHES} patches applied — check [WARN]/[FAIL] messages above")
        return False

    with open(filepath, "wb") as f:
        f.write(content)
    print(f"  [PASS] {patches_applied} patches applied")
    return True


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <path_to_index.js>")
        sys.exit(1)

    success = patch_cowork_linux(sys.argv[1])
    sys.exit(0 if success else 1)

File: patches/fix_cowork_linux.nim
# @patch-target: app.asar.contents/.vite/build/index.js
# @patch-type: nim
#
# Nim port of fix_cowork_linux.py — produces byte-identical output.
#
# The MODE_PREAMBLE_JS snippet lives in js/cowork_mode_preamble.js and is
# shared verbatim between the Python and Nim implementations. This Nim
# source embeds it at compile time via staticRead; the Python source reads
# it at runtime. No codegen; the .js file is the single source of truth.
#
# 12 sub-patches:
#   0  mode preamble at "use strict"; start
#   A  VM client loader (extends Li check to include Linux)
#   B  socket path (mode-selected Unix socket)
#   C1 bundle config (inject linux:{x64:[]})
#   C2 bundle lookup alias (runtime-gated linux->win32)
#   D  pathToClaudeCodeExecutable (dynamic on Linux)
#   E  error detection (extend /usr/local/bin/claude check)
#   F  present_files (allow host outputs dir)
#   G  smol-bin copy gate (runtime-gated on kvm mode)
#   H  event subscription guard (skip ZHe if socket absent, lazy 60s retry)
#
# Patches always run on clean bundles — no "already patched" fast path.

import std/[os, strformat, strutils, options]
import std/nre

const MODE_PREAMBLE_JS = staticRead("../js/cowork_mode_preamble.js")

proc replaceFirst(
    content: var string, pattern: Regex, subFn: proc(m: RegexMatch): string
): int =
  let maybe = content.find(pattern)
  if maybe.isNone:
    return 0
  let m = maybe.get()
  let bounds = m.matchBounds
  content = content[0 ..< bounds.a] & subFn(m) & content[bounds.b + 1 .. ^1]
  return 1

proc replaceAllRegex(
    content: var string, pattern: Regex, subFn: proc(m: RegexMatch): string
): int =
  var count = 0
  content = content.replace(
    pattern,
    proc(m: RegexMatch): string =
      inc count
      subFn(m),
  )
  return count

proc apply*(input: string): string =
  var content = input
  let original = input
  var patchesApplied = 0
  const EXPECTED_PATCHES = 10

  # ── Patch 0: Mode preamble at file start (right after "use strict";) ──
  block:
    const strictMarker = "\"use strict\";"
    let strictIdx = content.find(strictMarker)
    if strictIdx == 0:
      let insertionPoint = strictMarker.len
      content =
        content[0 ..< insertionPoint] & MODE_PREAMBLE_JS & content[insertionPoint .. ^1]
      echo "  [OK] 0 mode preamble: injected after \"use strict\""
      inc patchesApplied
    else:
      echo "  [FAIL] 0 mode preamble: \"use strict\"; not at file start"

  # ── Patch A: VM client loader — extend the install gate to include Linux ──
  #
  # History:
  #   <=v1.12603: `zo?IM={vm:L7i}:IM=(await import("@ant/claude-swift")).default`
  #     — a win32(zo) ternary: win32 loaded the TS vmClient `{vm:...}`, else
  #     imported the native `@ant/claude-swift`. We OR'd Linux into the win32
  #     side so Linux used the TS vmClient.
  #   v1.13576.0: REFACTORED. The ternary is gone. The loader is now
  #     `function d_t(){return tu()?sG||WoA||(...,WoA=(async()=>{try{return
  #     sG={vm:qti},...})()):null}` — gated entirely on tu() (MSIX/appPath
  #     install detection), with NO @ant/claude-swift fallback branch. tu()
  #     is false on Linux (no windowsStore, no "windowsapps" in execPath), so
  #     d_t() returns null and Cowork never gets a VM client.
  #
  # We widen the install gate to also fire on Linux, so d_t() loads the TS
  # vmClient (`{vm:qti}`) there. Anchored on the unique "[VM] Loading %s
  # module..." + "vmClient (TypeScript)" + `{vm:...}` body so we hit only the
  # loader and capture the minified gate-fn call (tu()).
  block:
    let pat =
      re"""(function [\w$]+\(\)\{return )([\w$]+\(\))(\?[\w$]+\|\|[\w$]+\|\|\([\w$]+\.info\("\[VM\] Loading %s module\.\.\.","vmClient \(TypeScript\)"\),[\w$]+=\(async\(\)=>\{try\{return [\w$]+=\{vm:[\w$]+\})"""
    let n = replaceAllRegex(
      content,
      pat,
      proc(m: RegexMatch): string =
        let prefix = m.captures[0]
        let gateCall = m.captures[1]
        let rest = m.captures[2]
        prefix & "(" & gateCall & "||process.platform===\"linux\")" & rest,
    )
    if n >= 1:
      echo &"  [OK] A VM client loader: install gate extended to Linux ({n} match)"
      inc patchesApplied
    else:
      echo "  [FAIL] A VM client loader: 0 matches"

  # ── Patch B: Socket path — mode-aware Unix socket on Linux ──
  block:
    # The original bytes pattern in Python is
    #   b'"\\\\\\\\.\\\\pipe\\\\cowork-vm-service"'
    # which decodes to the literal JS source bytes:
    #   "\\\\.\\pipe\\cowork-vm-service"
    # i.e. 8 backslashes encoding the Windows pipe path.
    const pipePath = "\"\\\\\\\\.\\\\pipe\\\\cowork-vm-service\""
    const pipeSearch = "=" & pipePath

    let idx = content.find(pipeSearch)
    if idx != -1:
      # Walk backwards to find start of variable name
      var start = idx - 1
      while start >= 0:
        let c = content[start]
        if c.isAlphaNumeric or c == '_' or c == '$':
          dec start
        else:
          break
      inc start
      let varName = content[start ..< idx]

      let replacement =
        varName & "=process.platform===\"linux\"?" &
        "(process.env.XDG_RUNTIME_DIR||\"/tmp\")+(globalThis.__coworkKvmMode?\"/cowork-kvm-service.sock\":\"/cowork-vm-service.sock\")" &
        ":" & pipePath

      content = content[0 ..< start] & replacement & content[idx + pipeSearch.len .. ^1]
      echo &"  [OK] B socket path: runtime-selected Unix socket on Linux (var={varName})"
      inc patchesApplied
    else:
      echo "  [FAIL] B socket path: pipe path not found"

  # ── Patch C1: Inject ",linux:{x64:[]}" into bundle files config ──
  block:
    const win32Marker = "win32:{"
    let win32Idx = content.find(win32Marker)
    if win32Idx >= 0:
      const x64Marker = ",x64:["
      let x64Idx = content.find(x64Marker, win32Idx)
      if x64Idx >= 0:
        # Position of the opening '['
        let arrayStart = x64Idx + x64Marker.len - 1
        var depth = 0
        var pos = arrayStart
        var matched = false
        while pos < content.len:
          let ch = content[pos]
          if ch == '[':
            inc depth
          elif ch == ']':
            dec depth
            if depth == 0:
              matched = true
              break
          inc pos
        if matched:
          let afterArray = pos + 1
          if afterArray < content.len and content[afterArray] == '}':
            const inject = ",linux:{x64:[]}"
            content = content[0 .. afterArray] & inject & content[afterArray + 1 .. ^1]
            echo "  [OK] C1 bundle config: linux platform added (empty file list)"
            inc patchesApplied
          else:
            echo "  [FAIL] C1 bundle config: unexpected structure after x64 array"
        else:
          echo "  [FAIL] C1 bundle config: x64 array not balanced"
      else:
        echo "  [FAIL] C1 bundle config: x64 array not found in win32 block"
    else:
      echo "  [FAIL] C1 bundle config: win32 block not found"

  # ── Patch C2: Alias linux→win32 at the VM-bundle lookup ──
  #
  # History:
  #   <=v1.12603: the bundle resolver indexed the file map by process.platform:
  #     `const A=process.platform,e=mSA();return Io.files[A]` (fns nX/fci). On
  #     Linux that yielded the (empty) linux entry, so we aliased linux→win32
  #     under kvm mode to reuse the win32 bundle list.
  #   v1.13576.0: REFACTORED. The platform-indexed resolver is gone. The only
  #     remaining VM-bundle lookup HARDCODES win32:
  #       function C3i(){...const A="win32",e=process.arch;...
  #         if(!((r=fo.files[A])!=null&&r[e]))return{...unsupported_architecture}
  #     i.e. on Linux the arch check already reads fo.files["win32"][arch] —
  #     exactly the linux→win32 alias this patch used to inject. The alias is
  #     therefore already upstream's behavior; there is no process.platform
  #     index left to rewrite.
  #
  # We first try the historical platform-indexed form; if it's absent, we
  # confirm the new hardcoded-win32 lookup exists (alias already satisfied)
  # and succeed without changes. Failing only if NEITHER is present guards
  # against a wholesale removal of the VM-bundle resolver.
  block:
    let pat =
      re"""(const ([\w$]+)=)process\.platform(,[\w$]+=[\w$]+\(\);return [\w$]+\.files\[\2\])"""
    let n = replaceAllRegex(
      content,
      pat,
      proc(m: RegexMatch): string =
        m.captures[0] &
          "(globalThis.__coworkKvmMode?(process.platform===\"linux\"?\"win32\":process.platform):process.platform)" &
          m.captures[2],
    )
    if n >= 1:
      echo &"  [OK] C2 bundle lookup alias: linux→win32 when kvm mode ({n} site(s))"
      inc patchesApplied
    else:
      # No platform-indexed lookup. Confirm the hardcoded-win32 lookup exists:
      #   const A="win32",e=process.arch; ... (i=<map>.files[A])!=null&&i[e]
      # The deref var (\2) is minified and changes every release (was `r`,
      # now `i`); capture it and backreference so we assert the real end-state.
      let hardcodedWin32Lookup = re(
        """const ([\w$]+)="win32",[\w$]+=process\.arch;[\s\S]{0,400}?\(([\w$]+)=[\w$]+\.files\[\1\]\)!=null&&\2\["""
      )
      if content.find(hardcodedWin32Lookup).isSome:
        echo "  [OK] C2 bundle lookup alias: lookup hardcoded to win32 upstream (linux→win32 alias already in effect; skipped)"
        inc patchesApplied
      else:
        echo "  [FAIL] C2 bundle lookup alias: no matching sites found (neither platform-indexed nor hardcoded-win32)"

  # ── Patch D: pathToClaudeCodeExecutable — dynamic on Linux ──
  block:
    const claudePathOld = "pathToClaudeCodeExecutable:\"/usr/local/bin/claude\""
    const claudePathNew =
      "pathToClaudeCodeExecutable:" &
      "(()=>{if(process.platform!==\"linux\")return\"/usr/local/bin/claude\";" &
      "const fs=require(\"fs\");" & "for(const p of[\"/usr/bin/claude\"," &
      "(process.env.HOME||\"\")+\"/.local/bin/claude\"," & "\"/usr/local/bin/claude\"])" &
      "if(fs.existsSync(p))return p;" &
      "try{return require(\"child_process\").execSync(\"which claude\",{encoding:\"utf-8\"}).trim()}" &
      "catch(e){}" & "return\"claude\"})()"

    let idx = content.find(claudePathOld)
    if idx != -1:
      content =
        content[0 ..< idx] & claudePathNew & content[idx + claudePathOld.len .. ^1]
      echo "  [OK] D Claude Code path: dynamic resolution on Linux"
      inc patchesApplied
    else:
      echo "  [FAIL] D Claude Code path: pattern not found"

  # ── Patch E: Error detection — extend Linux paths ──
  block:
    let pat = re"""([\w$]+)(\.includes\("/usr/local/bin/claude"\))"""
    let n = replaceFirst(
      content,
      pat,
      proc(m: RegexMatch): string =
        let v = m.captures[0]
        "(" & v & ".includes(\"/usr/local/bin/claude\")||" & v &
          ".includes(\"/usr/bin/claude\")||" & v & ".includes(\"/.local/bin/claude\"))",
    )
    if n >= 1:
      echo "  [OK] E error detection: extended for Linux paths"
      inc patchesApplied
    else:
      echo "  [FAIL] E error detection: pattern not found"

  # ── Patch F: present_files — allow host outputs dir paths ──
  block:
    let pat =
      re"""for\(const\{file_path:([\w$]+),vmPath:([\w$]+)\}of ([\w$]+)\)\{if\(([\w$]+)\(\2,([\w$]+)\.vmProcessName\)\)continue;\(([\w$]+)\?([\w$]+)\(\2,\6\):null\)===null&&([\w$]+)\.push\(\1\)\}"""
    let n = replaceAllRegex(
      content,
      pat,
      proc(m: RegexMatch): string =
        let fVar = m.captures[0]
        let pVar = m.captures[1]
        let lVar = m.captures[2]
        let scratchpadFn = m.captures[3]
        let tVar = m.captures[4]
        let cVar = m.captures[5]
        let resolveFn = m.captures[6]
        let uVar = m.captures[7]
        "for(const{file_path:" & fVar & ",vmPath:" & pVar & "}of " & lVar & "){" & "if(" &
          scratchpadFn & "(" & pVar & "," & tVar & ".vmProcessName))continue;" & "(" &
          cVar & "?" & resolveFn & "(" & pVar & "," & cVar & "):null)===null&&" &
          "(()=>{const _ho=" & tVar & ".getHostOutputsDir();" & "if(_ho&&(" & fVar &
          "===_ho||" & fVar & ".startsWith(_ho+\"/\")))return;" & uVar & ".push(" & fVar &
          ")})()}",
    )
    if n >= 1:
      echo &"  [OK] F present_files: host outputs dir allowed ({n} match)"
      inc patchesApplied
    else:
      echo "  [FAIL] F present_files: pattern not found"

  # ── Patch G: smol-bin copy gate — runtime-gated on kvm mode ──
  block:
    let pat =
      re"""if\(process\.platform==="win32"\)(\{const [\w$]+=[\w$]+\(\),[\w$]+=[\w$]+\.join\(process\.resourcesPath,`smol-bin\.)"""
    let n = replaceAllRegex(
      content,
      pat,
      proc(m: RegexMatch): string =
        "if(process.platform===\"win32\"||process.platform===\"linux\"&&globalThis.__coworkKvmMode)" &
          m.captures[0],
    )
    if n >= 1:
      echo &"  [OK] G smol-bin copy gate: kvm-mode Linux opt-in ({n} match)"
      inc patchesApplied
    else:
      echo "  [FAIL] G smol-bin copy gate: win32 gate pattern not found"

  # ── Patch H: Guard event subscription — skip if socket absent ──
  block:
    # Two known shapes (minified names change every release):
    #   Sync:  function ZHe(){if(qD)return;const e=PtA.createConnection(lbA)
    #   Async: function Zje(){return vy?Promise.resolve():new Promise((e,A)=>{const t=PoA.createConnection(xYA)
    let patSync =
      re"""function ([\w$]+)\(\)\{if\(([\w$]+)\)return;const ([\w$]+)=([\w$]+)\.createConnection\(([\w$]+)\)"""
    let patAsync =
      re"""function ([\w$]+)\(\)\{return ([\w$]+)\?Promise\.resolve\(\):new Promise\(\(([\w$]+),([\w$]+)\)=>\{const ([\w$]+)=([\w$]+)\.createConnection\(([\w$]+)\)"""

    proc buildGuardBody(funcName, pathVar: string): string =
      "if(process.platform===\"linux\"){" & "var _fs=require(\"fs\");" &
        "try{_fs.accessSync(" & pathVar & ");}catch(_e){" &
        "if(!globalThis.__coworkRetryLogged){" & "globalThis.__coworkRetryLogged=true;" &
        "console.log(\"[vm-client] Cowork socket not found at \"+" & pathVar &
        "+\", event subscription deferred\");" & "}" &
        "if(!globalThis.__coworkRetryTimer){" &
        "globalThis.__coworkRetryTimer=setInterval(function(){" & "try{" &
        "require(\"fs\").accessSync(" & pathVar & ");" &
        "console.log(\"[vm-client] Cowork socket detected at \"+" & pathVar &
        "+\", connecting...\");" & "clearInterval(globalThis.__coworkRetryTimer);" &
        "globalThis.__coworkRetryTimer=null;" &
        "globalThis.__coworkServiceAvailable=true;" & funcName & "();" & "}catch(_e2){}" &
        "},60000);" & "}"

    var matched = false

    let syncMatch = content.find(patSync)
    if syncMatch.isSome:
      let m = syncMatch.get()
      let funcName = m.captures[0]
      let connVar = m.captures[1]
      let newVar = m.captures[2]
      let netMod = m.captures[3]
      let pathVar = m.captures[4]

      let guard =
        "function " & funcName & "(){" & buildGuardBody(funcName, pathVar) & "return;" &
        "}" & "globalThis.__coworkServiceAvailable=true;" & "}" & "if(" & connVar &
        ")return;const " & newVar & "=" & netMod & ".createConnection(" & pathVar & ")"

      let bounds = m.matchBounds
      content = content[0 ..< bounds.a] & guard & content[bounds.b + 1 .. ^1]
      echo &"  [OK] H event subscription guard: socket check in {funcName} (sync, pathVar={pathVar})"
      matched = true
      inc patchesApplied

    if not matched:
      let asyncMatch = content.find(patAsync)
      if asyncMatch.isSome:
        let m = asyncMatch.get()
        let funcName = m.captures[0]
        let connVar = m.captures[1]
        let resolveVar = m.captures[2]
        let rejectVar = m.captures[3]
        let newVar = m.captures[4]
        let netMod = m.captures[5]
        let pathVar = m.captures[6]

        let guard =
          "function " & funcName & "(){" & buildGuardBody(funcName, pathVar) &
          "return Promise.resolve();" & "}" & "globalThis.__coworkServiceAvailable=true;" &
          "}" & "return " & connVar & "?Promise.resolve():new Promise((" & resolveVar &
          "," & rejectVar & ")=>{const " & newVar & "=" & netMod & ".createConnection(" &
          pathVar & ")"

        let bounds = m.matchBounds
        content = content[0 ..< bounds.a] & guard & content[bounds.b + 1 .. ^1]
        echo &"  [OK] H event subscription guard: socket check in {funcName} (async, pathVar={pathVar})"
        matched = true
        inc patchesApplied

    if not matched:
      echo "  [FAIL] H event subscription guard: neither sync nor async pattern found"

  # ── Checks ────────────────────────────────────────────────────
  if patchesApplied < EXPECTED_PATCHES:
    raise newException(
      ValueError,
      &"Only {patchesApplied}/{EXPECTED_PATCHES} patches applied — check [WARN]/[FAIL] messages above",
    )

  echo &"  [PASS] {patchesApplied} patches applied"
  result = content
  discard original

when isMainModule:
  let file = paramStr(1)
  echo "=== Patch: fix_cowork_linux ==="
  echo &"  Target: {file}"
  let input = readFile(file)
  let output = apply(input)
  if output != input:
    writeFile(file, output)
